### PR TITLE
smartplugin: add pause_item (instead of suspend item)

### PR DESCRIPTION
**Pull request by Morg42**
This item more or less replaces the suspend_item, but it hard-stops and re-runs the plugin instead of trying to suspend it while being active.

Modified and adjusten SmartPlugin and MqttPlugin as well as both sample plugins and their plugin.yaml files. 

Works with adjusted zigbee2mqtt Plugin; plugin wishing to implement this need to
- be restartable (though metadata restartable is not queried)
- implement code snippets in init, run, stop, parse_item, update_item (see sample plugins)

Also cleaned up SmartPlugin, commented the reason for using class members and added some type hinting.

As far as I can see, the changes should be non-breaking.

### CODE REVIEW

File: dev/sample_mqttplugin/__init__.py [hunk 1]
```diff
@@ -104,6 +119,13 @@ def parse_item(self, item):
                         with the item, caller, source and dest as arguments and in case of the knx plugin the value
                         can be sent to the knx with a knx write function within the knx plugin.
         """
+        # check for alive item
+        if item.property.path == self._pause_item_path:
+            self.logger.debug(f'alive item {item.property.path} registered')
```
Comment (onkelandy) [inline dev/sample_mqttplugin/__init__.py:124]: Should this rather be calles pause item here, too..?

File: dev/sample_mqttplugin/__init__.py [hunk 2]
```diff
@@ -88,6 +96,13 @@ def stop(self):
         self.logger.dbghigh(self.translate("Methode '{method}' aufgerufen", {'method': 'stop()'}))
         self.alive = False
 
+        # let the plugin change the state of pause_item
+        if self._pause_item:
+            self._pause_item(True, self.get_fullname())
+
+        # if you use schedulers, this stops all schedulers the plugin has started.
+        #self.scheduler_remove_all()
```
Comment (onkelandy) [inline dev/sample_mqttplugin/__init__.py:104]: I'd uncomment that cause there is no negative effect if the method is called even if no schedulers are created..?

File: dev/sample_mqttplugin/__init__.py [hunk 3]
```diff
@@ -78,6 +82,10 @@ def run(self):
 
         self.alive = True
 
+        # let the plugin change the state of pause_item
+        if self._pause_item:
```
Comment (onkelandy) [inline dev/sample_mqttplugin/__init__.py:86]: pause_item or pause_item_path..?

File: dev/sample_mqttplugin/__init__.py [hunk 4]
```diff
@@ -59,6 +59,10 @@ def __init__(self, sh):
         # Call init code of parent class (MqttPlugin)
         super().__init__()
 
+        # if you want to use an item to toggle plugin execution, enable the
+        # definition in plugin.yaml and uncomment the following line
+        #self._pause_item_path = self.get_parameter_value('pause_item')
```
Comment (onkelandy) [inline dev/sample_mqttplugin/__init__.py:64]: Should there be a check if the item exists..?

File: lib/model/smartplugin.py [hunk 5]
```diff
@@ -21,18 +21,18 @@
...
 from lib.utils import Utils
 from lib.translation import translate as lib_translate
+
 import logging
 import os
 import threading
 import asyncio
 import time
-import inspect
+from typing import Coroutine, Any
```
Comment (onkelandy) [inline lib/model/smartplugin.py:35]: The changes in this file are much more than just "add pause_item" as the title of this PR suggests.

File: dev/sample_mqttplugin/__init__.py (hunk 1 above)
Comment (Morg42) [reply dev/sample_mqttplugin/__init__.py]: yes.

File: dev/sample_mqttplugin/__init__.py (hunk 2 above)
Comment (Morg42) [reply dev/sample_mqttplugin/__init__.py:104]: As the sample mqtt plugin has no schedulers and I didn't want to add "default complexity" to the sample plugin, I kept this as a comment.

File: dev/sample_mqttplugin/__init__.py (hunk 3 above)
Comment (Morg42) [reply dev/sample_mqttplugin/__init__.py:86]: pause_item, as the item needs to be accessible for it to be changed in the next line.

File: dev/sample_mqttplugin/__init__.py (hunk 4 above)
Comment (Morg42) [reply dev/sample_mqttplugin/__init__.py:64]: Not necessarily; this never throws an error (but doesn't alert if the item is not present). As items are only initialized after plugin loading, testing for an existing item is not feasible here.

File: lib/model/smartplugin.py (hunk 5 above)
Comment (Morg42) [reply lib/model/smartplugin.py:35]: Hehe... no, this is the "cleanup". @msinn added the asyncio stuff at the bottom and I reordered much of the code to make public and "need to implement"-methods more visible when reading the source code. Asyncio code was added in commit ff7f747.

**Comment by Morg42** [comment]
On merging, also merge https://github.com/smarthomeNG/plugins/pull/945 to mitigate breaking changes to plugins using the old suspend_item mechanics.

**Pull request merged**